### PR TITLE
Handle sequence resets in anomaly predictions

### LINKED ISSUE #1954: Anomaly likelihood is reset on sequence reset?
Anomaly has internal state when using 
* Moving average (maybe this one could stay over seq. reset (?))
* likelihood (= likelihood/weighted modes)

- how are these treated at sequence reset? (eg sensor with `r`) 
@subutai

### PR DESCRIPTION
**Pull request by breznak**
Fixes #1954 
Introduces `reset()` and `isReady()` calls for `Anomaly, AnomalyLikelihood and MovingAverage`

TODO:
- [x] add OPF param setting for `CLAModel.resetSequenceStates(resetAnomaly=False)`

### CODE REVIEW

File: nupic/algorithms/anomaly.py [hunk 1]
```diff
@@ -154,9 +161,33 @@ def compute(self, activeColumns, predictedColumns,
       else:
         score = 0.0
 
+    self._iteration += 1
+    if not self.isReady():
+      score = 0.5
```
Comment (breznak) [inline nupic/algorithms/anomaly.py]: I think this makes sense. Maybe in CLAModel we can read this `burnin` value from TP and apply it to Anomaly.

File: nupic/algorithms/anomaly_likelihood.py [hunk 2]
```diff
@@ -130,31 +132,41 @@ def anomalyProbability(self, value, anomalyScore, timestamp=None):
...
+    """
+    return (len(self._historicalScores) >= self._probationaryPeriod and
+            self._iteration >= self._burninIterationStep)
+
+
+  def reset(self):
+    """
+    reset internal state after sequence reset
+    """
+    self._burninIterationStep = self._iteration + self._probationaryPeriod
```
Comment (breznak) [inline nupic/algorithms/anomaly_likelihood.py]: is this correct? First I was thinking of clearing `_historicalScores` and `_distribution` but that does not make sense, right?

File: nupic/frameworks/opf/clamodel.py [hunk 3]
```diff
@@ -235,19 +235,18 @@ def getParameter(self, paramName):
         (paramName))
 
 
-  def resetSequenceStates(self):
+  def resetSequenceStates(self, resetAnomaly=False):
```
Comment (breznak) [inline nupic/frameworks/opf/clamodel.py]: TODO: this should probably be set through some OPF description.py param.

File: nupic/utils.py [hunk 4]
```diff
@@ -81,6 +81,20 @@ def next(self, newValue):
...
+    self.slidingWindow = []
+    self.total = 0.0
+
+
+  def isReady(self):
+    """
+    Use after init and sequence reset() calls.
+    @return boolean - moving average is accurate if buffer is full.
+    """
+    return len(self.slidingWindow)==self.windowSize
```
Comment (breznak) [inline nupic/utils.py]: Is this correct, or moving average makes sense even for first few values?

File: tests/unit/nupic/algorithms/anomaly_test.py [hunk 5]
```diff
@@ -59,38 +59,38 @@ def testComputeRawAnomalyScorePartialMatch(self):
...
-    anomalyComputer = anomaly.Anomaly()
+    anomalyComputer = anomaly.Anomaly(claBurnInPeriod=0) # disable the burn-in period for the tests
     score = anomalyComputer.compute(array([2, 3, 6]), array([3, 5, 7]))
     self.assertAlmostEqual(score, 2.0 / 3.0)
 
 
   def testAnomalyCumulative(self):
     """Test cumulative anomaly scores."""
-    anomalyComputer = anomaly.Anomaly(slidingWindowSize=3)
+    anomalyComputer = anomaly.Anomaly(slidingWindowSize=3, claBurnInPeriod=0)
```
Comment (breznak) [inline tests/unit/nupic/algorithms/anomaly_test.py]: or should `0` here mean: "completely ignore :fire: in" ? and always return `isReady()` True?

**Comment by breznak** [comment]
@rhyolight can you please review code and the questions highlighted?

**Comment by breznak** [comment]
need to fix
tests/integration/nupic/opf/opf_experiment_results_test.py 
and maybe more

File: nupic/algorithms/anomaly.py [hunk 6]
```diff
@@ -72,7 +72,11 @@ class Anomaly(object):
   _supportedModes = (MODE_PURE, MODE_LIKELIHOOD, MODE_WEIGHTED)
 
 
-  def __init__(self, slidingWindowSize=None, mode=MODE_PURE, binaryAnomalyThreshold=None):
+  def __init__(self, 
+               slidingWindowSize=None,
+               mode=MODE_PURE,
+               binaryAnomalyThreshold=None,
+               claBurnInPeriod=300):
```
Comment (rhyolight) [inline nupic/algorithms/anomaly.py]: So I can specify `claBurnInPeriod=0` and get the previous behavior?

File: nupic/algorithms/anomaly.py (hunk 1 above)
Comment (rhyolight) [inline nupic/algorithms/anomaly.py]: :+1: for same behavior as anomaly likelihood.

File: nupic/algorithms/anomaly.py [hunk 7]
```diff
@@ -154,9 +161,33 @@ def compute(self, activeColumns, predictedColumns,
...
+
+
+  def isReady(self):
+    """
+    @return True when anomaly predictions should be accurate (after burn-in)
+    """
+    bMA = (self._movingAverage is None) or self._movingAverage.isReady()
+    bLike = (self._likelihood is None) or self._likelihood.isReady()
+    bCla = self._iteration > self._burnIn
+    return bCla and bMA and bLike
```
Comment (rhyolight) [inline nupic/algorithms/anomaly.py]: What about `len(self._historicalScores) < self._probationaryPeriod` as seen before in https://github.com/numenta/nupic/pull/2267/files#diff-afb5fcf88cb59a3944b9820b5ef3ca5dL133 ?

**Comment by rhyolight** [comment]
@scottpurdy I had some input, but someone versed in the anomaly algorithm should review.

File: nupic/algorithms/anomaly.py (hunk 6 above)
Comment (breznak) [inline nupic/algorithms/anomaly.py]: see https://github.com/numenta/nupic/pull/2267#discussion_r33456034 
I think this makes sense: 
- [ ] propagate setting of `claBurnInPeriod` from `Anomaly` to respective instances of MovingAverage, Likelihood. 
- [ ] on `claBurnIn = 0` `isReady()` would return always True, that way you get 100% prev. behavior.

**Comment by breznak** [comment]
additions: 
- `Anomaly.claBurnInPeriod` 0 will compute numbers right off, ignoring burnIn time.
- `Anomaly.claBurnInPeriod` None will not call reset(), ignoring 'r'/reset signal completely (current state)

None: 
- passing burnIn period to Likelihood is not 100%, as it uses another value (numEstimationSamples) and adds these 2 together (claBurnIn + numSamples) to get "ready"

File: nupic/algorithms/anomaly.py (hunk 1 above)
Comment (breznak) [inline nupic/algorithms/anomaly.py]: ok TP.burnIn is 2, which does not make sense for any stable learning. Should we bother raising the default to (300 as in likelihood?), or just ignore this idea?

File: nupic/algorithms/anomaly.py (hunk 6 above)
Comment (breznak) [inline nupic/algorithms/anomaly.py]: @rhyolight 
> So I can specify claBurnInPeriod=0 and get the previous behavior?

yes, but `None`, see doc for diff `0/None`

File: nupic/frameworks/opf/clamodel.py (hunk 3 above)
Comment (breznak) [inline nupic/frameworks/opf/clamodel.py]: OPF `anomalyParams['claBunrInPeriod'] >= 0` resets anomaly, None to ignore (default)

File: tests/unit/nupic/algorithms/anomaly_test.py (hunk 5 above)
Comment (breznak) [inline tests/unit/nupic/algorithms/anomaly_test.py]: None means that ^^

**Comment by breznak** [comment]
closing for more discussion in the issue!

**Pull request closed without merging**